share/getters/shrex.go

```diff
@@ -0,0 +1,79 @@
+package getters
+
+import (
+	"context"
+	"errors"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/p2p"
+	"github.com/celestiaorg/celestia-node/share/p2p/peers"
+	"github.com/celestiaorg/celestia-node/share/p2p/shrexeds"
+	"github.com/celestiaorg/celestia-node/share/p2p/shrexnd"
+
+	"github.com/celestiaorg/nmt/namespace"
+	"github.com/celestiaorg/rsmt2d"
+)
+
+var _ share.Getter = (*ShrexGetter)(nil)
+
+// ShrexGetter is a share.Getter that uses the shrex/eds and shrex/nd protocol to retrieve shares.
+type ShrexGetter struct {
+	edsClient *shrexeds.Client
+	ndClient  *shrexnd.Client
+
+	peerManager *peers.Manager
+}
+
+func (sg *ShrexGetter) GetShare(ctx context.Context, root *share.Root, row, col int) (share.Share, error) {
+	return nil, errors.New("shrex-getter: GetShare is not supported")
+}
+
+func (sg *ShrexGetter) GetEDS(ctx context.Context, root *share.Root) (*rsmt2d.ExtendedDataSquare, error) {
+	for {
+		peer, setStatus, err := sg.peerManager.Peer(ctx, root.Hash())
+		if err != nil {
+			return nil, err
+		}
+
+		eds, err := sg.edsClient.RequestEDS(ctx, root.Hash(), peer)
+		switch err {
+		case nil:
+			setStatus(peers.ResultSuccess)
+			return eds, nil
+		case p2p.ErrInvalidResponse:
+			setStatus(peers.ResultPeerMisbehaved)
+		case context.Canceled, context.DeadlineExceeded:
+			setStatus(peers.ResultFail)
```

Comment:
No need to set result here.

Reply:
I tied to point to contex errors cases. If we introduce backoff no need to put on backoff if it is context.Deadline